File: LeetCode/Python/Easy/1582-Special-Positions-in-a-Binary-Matrix/solution.py
class Solution:
    def numSpecial(self, mat: List[List[int]]) -> int:
        result=0

        rows=len(mat)

        columns=len(mat[0])

        one_positions=list()

        for row in range(0, rows):
            for column in range(0, columns):
                if mat[row][column]:
                    one_positions.append((row, column))

        for (row, column) in one_positions:
            if self.single_one_in_row(mat, columns, row, column) and self.single_one_in_column(mat, rows, row, column):
                result+=1

        return result

    def single_one_in_row(self, mat: List[List[int]], columns: int, row: int, column: int) -> bool:
        for c in range(0, column):
            if mat[row][c]:
                return False

        for c in range(column + 1, columns):
            if mat[row][c]:
                return False

        return True

    def single_one_in_column(self, mat: List[List[int]], rows: int, row: int, column: int) -> bool:
        for r in range(0, row):
            if mat[r][column]:
                return False

        for r in range(row + 1, rows):
            if mat[r][column]:
                return False

        return True

File: LeetCode/Rust/Easy/1582-Special-Positions-in-a-Binary-Matrix/solution.rs
impl Solution
{
    pub fn num_special(mat: Vec<Vec<i32>>) -> i32
    {
        let mut result: i32=0;

        let rows: usize=mat.len();

        let columns: usize=mat[0].len();

        let mut one_positions: Vec<(usize, usize)>=Vec::new();

        for row in 0..rows
        {
            for column in 0..columns
            {
                if mat[row][column]==1
                {
                    let position: (usize, usize)=(row, column);
                    
                    one_positions.push(position);
                }
            }
        }

        for (row, column) in one_positions.iter()
        {
            let value: usize=1;
            
            if Self::single_one_in_row(&mat, columns, *row, *column, value) && Self::single_one_in_column(&mat, rows, *row, *column, value)
            {
                result+=1;
            }
        }

        result
    }

    pub fn single_one_in_row(mat: &Vec<Vec<i32>>, columns: usize, row: usize, column: usize, value: usize) -> bool
    {
        for c in 0..column
        {
            if mat[row][c]==(value as i32)
            {
                return false;
            }
        }

        for c in column + 1..columns
        {
            if mat[row][c]==(value as i32)
            {
                return false;
            }
        }

        true
    }

    pub fn single_one_in_column(mat: &Vec<Vec<i32>>, rows: usize, row: usize, column: usize, value: usize) -> bool
    {
        for r in 0..row
        {
            if mat[r][column]==(value as i32)
            {
                return false;
            }
        }

        for r in row + 1..rows
        {
            if mat[r][column]==(value as i32)
            {
                return false;
            }
        }

        true
    }
}